_posts/2021-03-13-understanding-and-using-rubys-powerful-grep-method.md

```diff
@@ -0,0 +1,178 @@
+---
+layout: post
+title:  "Understanding and using Ruby's powerful #grep method"
+date:   2021-03-13 12:00:00 +0100
+author: "Shadi Rezek"
+categories: techtips ruby
+og_image: "/images/blog/posts/understanding-and-using-rubys-powerful-grep-method/grep.png"
+---
+
+Ruby is known for its many handy methods. I'd like to take a look at one that is a little bit less known, but quite powerful nonetheless: the `#grep` method.
+
+<img src="/images/blog/posts/understanding-and-using-rubys-powerful-grep-method/grep.png" style="width:300px;margin-left: 15px;float:right;" alt="#grep" />
+
+## About the method
+
+[#grep](https://ruby-doc.org/core-3.0.0/Enumerable.html#method-i-grep){:target="_blank"} is one of many instance methods defined on the `Enumerable` module. This essentially means that you can use it on all Ruby classes that include this module, e.g. `Array` or `Hash`. The `Enumerable` module is the place where many other popular "collection" methods are defined - `#map`, `#select`, and the like. Whatever Ruby object you can use the `#map` on, you should be able to use the `#grep` on as well.
+
+## "Getting a #grep"
+
+Let's see an example usage of the `#grep` method:
+
+{% highlight ruby %}
+array = ['foo', 'bar', 'baz']
+
+array.grep(/^ba/)
+# => ["bar", "baz"]
+{% endhighlight %}
+
+Seems pretty handy, right? And that's just one way of using it. But before showing more examples, let's deconstruct this to understand exactly how it works.
+
+What happens under the hood here could be mimicked with a code like this:
+
+{% highlight ruby %}
+array.select { |element| /regexp/ === element }
+{% endhighlight %}
+
+Or, more generally:
+
+{% highlight ruby %}
+collection.select { |element| pattern === element }
+{% endhighlight %}
+
+Notice a few things here:
+- The return value will always be an array. The length of this array will be between 0 (empty), and the size of the original collection (which is exactly how `#select` behaves too).
+- We are using the _triple equals_ (`#===`) method ("operator").
+  The `#===` is probably worthy of a separate article. But for now, we can just recall that it's implemented in classes as sort of a "lighter" (meaning: less strict) equality method (also called "case" equality due to its usage in case statements). We can illustrate the difference between regular and "case" equality with an example:
```

Comment:
I'm a bit confused. 

`#===` is more strict than `#==` I believe.
If that's the case then `classes as sort of a "lighter" (meaning: less strict) equality method` could be changed to `classes as sort of a "stronger" (meaning: more strict) equality method`

Reply:
I might be wrong and it's possible I don't get Ruby.

```
(1..2) == 2 #=> false
(1..2) === 2 #=> true
```

Above example is also a bit confusing for me. I don't know why `(1..2) === 2` is true. Is it because a range has 2 elements?

Reply:
> 
> `#===` is more strict than `#==` I believe.
Actually, it's the other way round in Ruby. The triple equals is a "case equality", which is the lighter of the two.

Reply:
> 
> Is it because a range has 2 elements?

It's because `2` falls within the range on `1..2`. This is how the `#===` method is implemented in the Range class.
This allows constructs like:
```ruby
case number
when (0..3) then 'low'
when (4..6) then 'medium'
when (7..10) then 'high'
end
```
(case statements use `===` method under the hood as well)

I'll think if there is a way to improve the description in the article.

Reply:
Thanks for the explanation. 

> Actually, it's the other way round in Ruby. The triple equals is a "case equality", which is the lighter of the two.

I didn't know about it. Probably I assumed === is stronger because of how it is in JS.